server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java

```diff
@@ -0,0 +1,362 @@
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import java.util.function.LongSupplier;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class PublicationTests extends ESTestCase {
+
+    class MockNode {
+
+        MockNode(Settings settings, DiscoveryNode localNode) {
+            this.localNode = localNode;
+            ClusterState initialState = CoordinationStateTests.clusterState(0L, 0L, localNode,
+                VotingConfiguration.EMPTY_CONFIG, VotingConfiguration.EMPTY_CONFIG, 0L);
+            coordinationState = new CoordinationState(settings, localNode, new CoordinationStateTests.InMemoryPersistedState(0L,
+                initialState));
+        }
+
+        final DiscoveryNode localNode;
+
+        final CoordinationState coordinationState;
+
+        public MockPublication publish(ClusterState clusterState, Discovery.AckListener ackListener, Set<DiscoveryNode> faultyNodes) {
+            PublishRequest publishRequest = coordinationState.handleClientValue(clusterState);
+            MockPublication currentPublication = new MockPublication(Settings.EMPTY, publishRequest, ackListener, () -> 0L) {
+                @Override
+                protected boolean isPublishQuorum(CoordinationState.VoteCollection votes) {
+                    return coordinationState.isPublishQuorum(votes);
+                }
+
+                @Override
+                protected Optional<ApplyCommitRequest> handlePublishResponse(DiscoveryNode sourceNode, PublishResponse publishResponse) {
+                    return coordinationState.handlePublishResponse(sourceNode, publishResponse);
+                }
+            };
+            currentPublication.start(faultyNodes);
+            return currentPublication;
+        }
+    }
+
+    abstract class MockPublication extends Publication {
+
+        final PublishRequest publishRequest;
+
+        ApplyCommitRequest applyCommit;
+
+        boolean completed;
+
+        boolean committed;
+
+        Map<DiscoveryNode, ActionListener<PublishWithJoinResponse>> pendingPublications = new HashMap<>();
+        Map<DiscoveryNode, ActionListener<TransportResponse.Empty>> pendingCommits = new HashMap<>();
+
+        MockPublication(Settings settings, PublishRequest publishRequest, Discovery.AckListener ackListener,
+                               LongSupplier currentTimeSupplier) {
+            super(settings, publishRequest, ackListener, currentTimeSupplier);
+            this.publishRequest = publishRequest;
+        }
+
+        @Override
+        protected void onCompletion(boolean committed) {
+            completed = true;
+            this.committed = committed;
+        }
+
+        @Override
+        protected void onPossibleJoin(DiscoveryNode sourceNode, PublishWithJoinResponse response) {
+
```

Comment:
We should assert that this is called (and I guess that we can say something about the response that is passed in).

Reply:
sure, done in f09659ceda6